scripts/rewrite-declation-statements-for-typescript.js

```diff
@@ -0,0 +1,214 @@
+
+  traverse(ast, {
+    ImportDeclaration(path) {
+      const packageStr = rewriteDeclarationSource(path.node, srcFile, rootDir);
+      const {value: sourceValue} = path.node.source;
+
+      const hasThirdPartyTransformed = sourceValue.includes(THIRD_PARTY_PATH);
+      if (sourceValue !== packageStr && !hasThirdPartyTransformed) {
+        const importDeclaration = t.importDeclaration(path.node.specifiers, t.stringLiteral(packageStr));
+        // Preserve comments above import statements, since this is most likely
+        // the license comment.
+        if (path.node.comments && path.node.comments.length > 0) {
+          for (let i = 0; i < path.node.comments.length; i++) {
+            const commentValue = path.node.comments[i].value;
+            importDeclaration.comments = importDeclaration.comments || [];
+            importDeclaration.comments.push({type: 'CommentBlock', value: commentValue});
+          }
+        }
+        path.replaceWith(importDeclaration);
+      }
+    },
+  });
+
+  const {code: outputCode} = recast.print(ast, {
+    objectCurlySpacing: false,
+    quote: 'single',
+    trailingComma: {
+      objects: false,
+      arrays: true,
+      parameters: false,
+    },
+  });
+
+  fs.writeFileSync(srcFile, outputCode, 'utf8');
+  logProgress(`[rewrite] ${srcFile}`);
+}
+
+function rewriteDeclarationSource(node, srcFile, rootDir) {
+  let source = node.source.value;
+  const pathParts = source.split('/');
+  const isMDCImport = pathParts[0] === '@material';
+
+  // format `source` for @material/foo imports
+  if (isMDCImport) {
+    const modName = pathParts[1]; // @material/<modName>
+    const atMaterialReplacementPath = `${rootDir}/mdc-${modName}`;
+    const rewrittenSource = [atMaterialReplacementPath].concat(pathParts.slice(2)).join('/');
+    source = rewrittenSource;
+  }
+
+  return patchNodeForDeclarationSource(source, srcFile, rootDir, node);
+}
+
+function isThirdPartyModule(source) {
+  const wouldLoadAsFileOrDir = ['./', '/', '../'].some((s) => source.indexOf(s) === 0);
+  return !wouldLoadAsFileOrDir;
+}
+
+function patchNodeForDeclarationSource(source, srcFile, rootDir, node) {
+  let resolvedSource = source;
+  const basedir = path.dirname(srcFile);
+  // See: https://nodejs.org/api/modules.html#modules_all_together (step 3)
+  if (isThirdPartyModule(source)) {
+    // for focus-trap
+    assert(source.indexOf('@material') < 0, '@material/* import sources should have already been rewritten');
+    patchDefaultImportIfNeeded(node);
+    resolvedSource = `${THIRD_PARTY_PATH}${camelCase(source)}`;
+    return resolvedSource;
+  } else {
+    const fileDir = resolve.sync(source, {
+      basedir,
+      extensions: ['.ts', '.js'],
+    });
+    const srcDirectory = path.dirname(srcFile);
+    resolvedSource = path.relative(srcDirectory, fileDir);
+
+    const packageStr = resolvedSource
+      .replace('.js', '')
+      .replace('.ts', '')
+      .replace('.index', '');
+    return `./${packageStr}`;
```

Comment:
I noticed there's an extra leading `./` on converted relative imports that doesn't need to be there in the output. The leading `./` here might be the cause. I'm not sure if this is necessary; if an import is already relative (and thus already has a leading `./` or `../`, would it even hit this path?

Reply:
I put it there, since I'm more used to seeing it like that. I don't think it is necessary. I can remove.

Reply:
Turns out I was wrong about this and it was also being used for imports within the same directory, so we had to put this back.